### ISSUE TITLE
Video Options : Inconsistent tracking of Full Screen and Resolutions

### ISSUE THREAD

**Issue by Dragon-Baroque** [exult/exult#383]
While working on Exult support of SDL3, I have been chasing an **Exult crash when switching Full Screen to Enabled, changing the Resolution, and trying Apply** in the Video Options gump.

I was in Shared mode ( Same settings for window and fullscreen ) YES, and this had the - actually surprising - effect of not changing the list of resolutions in the Window Size / Display Mode button.

But the `VideoOptions_gump::toggle_resolution` IS full screen dependent and, in my case, tried to apply the button position in the windowed resolution list to the full screen resolution list. But is SDL3 - I have to understand why - the full screen resolution list is limited to the max resolution of my display, plus the current display resolution. Two entries instead of the seventeen or so of the windowed resolution list. Thus the `toggle_resolution` index was way outside the array range, setting the resolution to a random value, often zero * zero. Later the zero resolution is used as a divider in `Image_window::get_draw_dims`, then crash.

I am confused about the meaning of the Shared mode YES in the Video Options gump. If it is NO, the list of Resolutions in the Window Size / Display Mode button is tracking the Full Screen option, and the crash does not occur.
Even if the Shared Mode is set to YES, shouldn't at least the Resolutions of the Window Size / Display Mode button be tracking the Full Screen mode ? Or should the `toggle_resolution` rely instead on the original Full Screen state from the entry to the Vide Options gump, not its current state ?

**Comment by Dragon-Baroque** [comment]
Strangely the Resolution button in the Video Option gump changes its name according to the Full Screen property. But not its list of values.
The Video Options gump has three buttons with dynamic values, the Resolution, the Scaling, and the Aspect Correction buttons.
There are only two triggers, one on the Scaler button - for the Scaling -, one on the Fill Mode button - for the Aspect Correction -. Should the Full Screen button be the third trigger, like this, replacing :
```
void toggle_fullscreen(int state) {
    if (share_settings)
        fullscreen = state;
    else {
        load_settings(state); // overwrites old settings
        rebuild_buttons();
    }
}
```
by
```
void toggle_fullscreen(int state) {
    if (share_settings) {
        fullscreen = state;
        rebuild_dynamic_buttons(); // rebuild the resolutions
    } else {
        load_settings(state); // overwrites old settings
        rebuild_buttons();
    }
}
```

**Comment by DominusExult** [comment]
I have to confess that I don't quite understand 🤷‍♂️

**Comment by Dragon-Baroque** [comment]
I am having problems with the switch to FullScreen mode in Exult. This append is quite long, please bear with me.

**Symptoms**
For you to reproduce : when I open Exult ( at Git master ) without an `.exult.cfg`, it opens as windowed, the Video Options gump displays
- at the top => Fullscreen: Disabled,
- second row => Window size: 1024x768, with possible values from 320x200 to 1920x1440,
- and at the bottom => Same settings for windowed and fullscreen : Yes.

Then I switch Fullscreen to Enabled. The second row button label changes to Display Mode but its value remains at 1024x768. **And its possible values remain at 320x200 .. 1920x1440**.

If I just switch Fullscreen to Enabled and then Apply the change, Exult switches to Fullscreen. I have no way to check whether the current Display Mode 1024x768 has any effect.
**1 =>** If I then close the Video Options gump with the red check mark, and reopen it immediately - no need to close the Main gump - then it displays
- at the top => Fullscreen: Enabled,
- second row => Display Mode: **1920x1080**, with possible values **a whole new set from 640x480 to 1920x1080**,
- and at the bottom => Same settings for windowed and fullscreen : Yes.

I can then change the Display Mode to my taste.

**2 =>** On the other hand, if I remain in the Video Options gump after Apply, and then change the resolution with the Display Mode button to the proposed 1920x1440, and then I press Apply, then Exult crashes with `Floating Point Exception`.
The same crash occurs if in the Video Options gump on a windowed Exult, I switch to Fullscreen Enabled and I change the Display Mode, and only then I press Apply.

Your experience may vary, depending on the number of possible fullscreen modes of your display. If for example, you own a larger display than mine, then the list of Display Modes obtained when you open the Video Options gump while in Fullscreen may be way longer, and the crash does not occur. 

On the other hand, it the crash does not occur, you may try the exact opposite process :
- Start from a blank Exult in windowed 1024x768, Switch it to FullScreen in the Video Options gump, Exit Exult.
- Open Exult ( in Fullsceen mode, now ), Open the Video Options gump, and
- Switch to Fullscreen Disabled, Change the Window size to the largest permitted size, and press Apply.

You could obtain the crash.

**Diagnosis**
It boils down to a mismatch in the Window Size / Display Mode button in what it displays and what it does when clicked :
- The Video Options gump maintains two lists of possible resolutions, one for the Windowed mode, one for the Fullscreen mode,
- The possible resolution values of the second button are kept at the list of resolutions of the Fullscreen vs Windowed state of the Video Options gump **when it was opened**, _because the Same settings for windowed and fullscreen is Yes_,
- But the button handler code indexes by the button list index, the list of resolutions of the **current state** of the Fullscreen button,
- Thus, if you change the Fullscreen mode and remain in the Video Options gump, the second button displays the resolutions from **one** of the two lists, but its click handler indexes the **other** list,
- Then if you choose the max resolution, its index in one list may be outside the other list, the resolution it then fetches is random memory, and if it is zero, then you get a `Floating Point Exception` later.

**Proposal**
Even with Same settings for windowed and fullscreen : Yes, the list of values of the second button should reflect the current state of the Fullscreen button, to make the possible values of the button consistent with its click handler.

**Comment**
The Video Options gump separates its own refresh actions into
- a static part in `rebuild_buttons` which is called at the opening of the gump, and by the Fullscreen button when Same settings for windowed and fullscreen is No,
- and a dynamic part in `rebuild_dynamic_buttons` which is called at the opening of the gump and also when you click on the Scaler or the Fill Mode buttons, regardless of the state of Same settings for windowed and fullscreen.

Then `rebuild_dynamic_buttons` refreshes the values of the Window Size / Display Mode button, of the Scaling button and of the AR Correction button.

My code proposal consists in calling the `rebuild_dynamic_buttons` in the Fullscreen button click handler when Same settings for windowed and fullscreen is Yes.

**Comment by DominusExult** [comment]
It's been a while :)
Going through the issues and I can indeed confirm the crash on macOS. I have no idea why I never followed up on this back then. Sorry.

**Comment by DominusExult** [comment]
do you still get the crash with SDL3

**Comment by Dragon-Baroque** [comment]
> Do you still get the crash with SDL3 ?

No longer, sorry to have left the issue open. What about you ?

> Going through the issues and I can indeed confirm the crash on macOS. 

Do you get the crash ?

**Comment by DominusExult** [comment]
Yes, it can be closed. I still got it with SDL2 but now no more.
